From ca3dc9ea4562e9363007f65db4807ca7056bcbb6 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Mon, 18 Aug 2014 19:56:34 -0400
Subject: [scsi] qla2xxx: Allow the next firmware dump if the previous dump capture fails for ISP8044

Message-id: <1408391817-6389-47-git-send-email-cdupuis@redhat.com>
Patchwork-id: 87933
O-Subject: [PATCH RHEL 7.1 e-stor 46/69] qla2xxx: Allow the next firmware dump if the previous dump capture fails for ISP8044.
Bugzilla: 1089346
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Saurav Kashyap <saurav.kashyap@qlogic.com>

scsi commit id edaa5c74177a6856f108b4909407985b6d3f951c

Signed-off-by: Giridhar Malavali <giridhar.malvali@qlogic.com>
Signed-off-by: Saurav Kashyap <saurav.kashyap@qlogic.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla2xxx/qla_def.h b/drivers/scsi/qla2xxx/qla_def.h
index 28d9351..d4046fa 100644
--- a/drivers/scsi/qla2xxx/qla_def.h
+++ b/drivers/scsi/qla2xxx/qla_def.h
@@ -3121,6 +3121,7 @@ struct qla_hw_data {
  uint32_t fw_dump_len;
  int  fw_dumped;
  int  fw_dump_reading;
+ int  prev_minidump_failed;
  dma_addr_t eft_dma;
  void  *eft;
 /* Current size of mctp dump is 0x086064 bytes */
diff --git a/drivers/scsi/qla2xxx/qla_nx.c b/drivers/scsi/qla2xxx/qla_nx.c
index 0f94752..a67f264 100644
--- a/drivers/scsi/qla2xxx/qla_nx.c
+++ b/drivers/scsi/qla2xxx/qla_nx.c
@@ -3111,18 +3111,18 @@ qla82xx_check_md_needed(scsi_qla_host_t *vha)
 
  if (ql2xmdenable) {
   if (!ha->fw_dumped) {
-   if (fw_major_version != ha->fw_major_version ||
+   if ((fw_major_version != ha->fw_major_version ||
        fw_minor_version != ha->fw_minor_version ||
-       fw_subminor_version != ha->fw_subminor_version) {
+       fw_subminor_version != ha->fw_subminor_version) ||
+       (ha->prev_minidump_failed)) {
     ql_dbg(ql_dbg_p3p, vha, 0xb02d,
-        "Firmware version differs "
-        "Previous version: %d:%d:%d - "
-        "New version: %d:%d:%d\n",
+        "Firmware version differs Previous version: %d:%d:%d - New version: %d:%d:%d, prev_minidump_failed: %d.\n",
         fw_major_version, fw_minor_version,
         fw_subminor_version,
         ha->fw_major_version,
         ha->fw_minor_version,
-        ha->fw_subminor_version);
+        ha->fw_subminor_version,
+        ha->prev_minidump_failed);
     /* Release MiniDump resources */
     qla82xx_md_free(vha);
     /* ALlocate MiniDump resources */
diff --git a/drivers/scsi/qla2xxx/qla_nx2.c b/drivers/scsi/qla2xxx/qla_nx2.c
index 9bea074..b34d2e4 100644
--- a/drivers/scsi/qla2xxx/qla_nx2.c
+++ b/drivers/scsi/qla2xxx/qla_nx2.c
@@ -3125,6 +3125,7 @@ skip_nxt_entry:
       "Dump data mismatch: Data collected: "
       "[0x%x], total_data_size:[0x%x]\n",
       data_collected, ha->md_dump_size);
+  rval = QLA_FUNCTION_FAILED;
   goto md_failed;
  }
 
@@ -3149,10 +3150,12 @@ qla8044_get_minidump(struct scsi_qla_host *vha)
 
  if (!qla8044_collect_md_data(vha)) {
   ha->fw_dumped = 1;
+  ha->prev_minidump_failed = 0;
  } else {
   ql_log(ql_log_fatal, vha, 0xb0db,
       "%s: Unable to collect minidump\n",
       __func__);
+  ha->prev_minidump_failed = 1;
  }
 }
 
diff --git a/drivers/scsi/qla2xxx/qla_os.c b/drivers/scsi/qla2xxx/qla_os.c
index 48827cf..cdcfdc8 100644
--- a/drivers/scsi/qla2xxx/qla_os.c
+++ b/drivers/scsi/qla2xxx/qla_os.c
@@ -2829,6 +2829,7 @@ skip_dpc:
 
  base_vha->flags.init_done = 1;
  base_vha->flags.online = 1;
+ ha->prev_minidump_failed = 0;
 
  ql_dbg(ql_dbg_init, base_vha, 0x00f2,
      "Init done and hba is online.\n");
-- 
1.7.1